libqtile/notify.py

```diff
@@ -110,22 +136,174 @@ def service(self):
+            # Assign prev_id to notifications indexing dict
+            self._notifications_idx[notif.id] = dict(
+                prev_id=prev_id,
+                next_id=None
+            )
+
+            # Set next_id of previous notification to this notification's id.
+            if prev_id is not None:
+                self._notifications_idx[prev_id]['next_id'] = notif.id
+
+            notifs_len = len(self._notifications)
+
+        for callback in self.callbacks_add:
+            # Don't let the subscriber crash us
+            try:
+                callback(notif)
+            except Exception as e:
+                logger.error(e)
+
+        return notifs_len
+
+    def delete(self, *notifs):
+        """Delete one or more notifications given the notifications or their
+        ids.
+        """
+        # Should not block the main thread
+        if main_thread() == current_thread():
+            Thread(target=self.delete, args=notifs).start()
+            return
+        notifs_deleted = []
+        with self._lock:
+            for notif in notifs:
+                if isinstance(notif, int):
+                    notif_id = notif
+                elif isinstance(notif, Notification):
+                    notif_id = notif.id
+                else:
+                    continue
+
+                notif = self._notifications.get(notif_id, None)
+                if notif is None:
+                    continue
+                notifs_deleted.append(notif)
+
+                # Get previous and next ids for re-indexing
+                notif_idx = self._notifications_idx[notif_id]
+                prev_id = notif_idx['prev_id']
+                next_id = notif_idx['next_id']
+
+                # Next notification should point to previous of deleted
+                if next_id is not None:
+                    self._notifications_idx[next_id]['prev_id'] = prev_id
+
+                # Previous notification should point to next of deleted
+                if prev_id is not None:
+                    self._notifications_idx[prev_id]['next_id'] = next_id
+
+                del self._notifications_idx[notif_id]
+                del self._notifications[notif_id]
+
+        # Do nothing if nothing was deleted
+        if len(notifs_deleted) == 0:
+            return
+
+        for callback in self.callbacks_delete:
+            # Don't let the subscriber crash us
+            try:
+                callback(*notifs_deleted)
+            except Exception as e:
+                logger.error(e)
+
+    def delete_all(self):
+        with self._lock:
+            notifs_deleted = self._notifications.values()
+            self._notifications = OrderedDict()
+            self._notifications_idx = {}
+
+        for callback in self.callbacks_delete:
+            # Don't let the subscriber crash us
+            try:
+                callback(*notifs_deleted)
+            except Exception as e:
+                logger.error(e)
```

Comment:
Could you change this to `logger.exception` so that the traceback is logged?

Reply:
Sure!